### PR TITLE
Feature/improve export performance

### PR DESCRIPTION
**Pull request by Alexander-Hjelm**

### CODE REVIEW

**Comment by Alexander-Hjelm** [comment]
Based on the work of: https://github.com/solidify/jira-azuredevops-migrator/pull/534

**Comment by Alexander-Hjelm** [comment]
@moerketh, fix the code smells :)

**Comment by Alexander-Hjelm** [comment]
@moerketh I ran my tests for the Agile demo project, but it seems like I'm always getting the following error for your branch:

```
Jira Exporter v0.0.1
Copyright (c) Solidify
Wow, there's a newer release out (v2.3.117)! We recommend downloading it for latest features and fixes.
[I][11:32:31] Connecting to Jira...
[I][11:32:31] Retrieving Jira fields...
[I][11:32:33] Retrieving Jira link types...
[I][11:32:33] Export started. Exporting 17 items.
[E][11:32:35] Unexpected migration error.
[System.ArgumentException] System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at JiraExport.JiraProvider.GetCustomId(String propertyName) in D:\a\1\s\jira-azuredevops-migrator\src\WorkItemMigrator\JiraExport\JiraProvider.cs:line 401
   at JiraExport.JiraCommandLine.ExecuteMigration(CommandOption user, CommandOption password, CommandOption url, CommandOption configFile, Boolean forceFresh, CommandOption continueOnCritical) in D:\a\1\s\jira-azuredevops-migrator\src\WorkItemMigrator\JiraExport\JiraCommandLine.cs:line 96: 
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at JiraExport.JiraProvider.GetCustomId(String propertyName) in D:\a\1\s\jira-azuredevops-migrator\src\WorkItemMigrator\JiraExport\JiraProvider.cs:line 401
   at JiraExport.JiraCommandLine.ExecuteMigration(CommandOption user, CommandOption password, CommandOption url, CommandOption configFile, Boolean forceFresh, CommandOption continueOnCritical) in D:\a\1\s\jira-azuredevops-migrator\src\WorkItemMigrator\JiraExport\JiraCommandLine.cs:line 96
[I][11:32:35] Export complete. Exported 17 items (1 errors, 0 warnings) in 00:00:03.
```

Any idea why that could be? If I run the test for master, it works fine

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs [hunk 1]
```diff
@@ -405,26 +395,24 @@ public string GetUserEmail(string usernameOrAccountId)
...
+            if (JiraNameFieldCache == null)
             {
-                var nameField = (JValue)item.SelectToken("name");
-                if (nameField.Value.ToString().ToLower() == propertyName.ToLower())
-                {
-                    var idField = (JValue)item.SelectToken("id");
-                    customId = idField.Value.ToString();
-                }
+                var response = (JArray)_jiraServiceWrapper.RestClient.ExecuteRequestAsync(Method.GET, $"{JiraApiV2}/field").Result;
+                JiraNameFieldCache = response
```
Comment (Alexander-Hjelm) [inline src/WorkItemMigrator/JiraExport/JiraProvider.cs:401]: The error I mentioned happend here

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs [hunk 2]
```diff
@@ -405,26 +395,24 @@ public string GetUserEmail(string usernameOrAccountId)
...
-                {
-                    var keyField = (JValue)item.SelectToken("key");
-                    if (keyField != null && keyField.Value != null
-                        && keyField.Value.ToString().ToLower() == propertyName.ToLower())
-                    {
-                        var idField = (JValue)item.SelectToken("id");
-                        customId = idField.Value.ToString();
-                    }
-                }
+            JiraNameFieldCache.TryGetValue(propertyName.ToLower(), out customId);
```
Comment (Alexander-Hjelm) [inline src/WorkItemMigrator/JiraExport/JiraProvider.cs:411]: @moerketh, what does this call do?

**Comment by moerketh** [comment]
So probably there are multiple fields with the same 'name' in your Jira instance. 

The previous code used the first match that it finds in the api response. That's an interesting case because it might not have a correct migration result for both fields. We should probably raise a separate issue for that. I suggest we create a test for this and allow the duplicate for now in the same way as the old code did.

**Comment by Alexander-Hjelm** [comment]
This would make most sense I think. Could you implement if and add a warning statement in case we encounter multiple fields with the same name?

If you don't have the time, lmk and I can have a crack at it as well

**Comment by moerketh** [comment]
Alright, I added a unittest for this behavior and changed the implementation to handle this with a warning log.

**Comment by moerketh** [comment]
I cannot update this branch directly, I created another pull request for this #539

**Comment by Alexander-Hjelm** [comment]
Now it looks like everything is running fine! I'll merge this as soon as the checks complete :)

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs [hunk 3]
```diff
@@ -405,25 +395,40 @@ public string GetUserEmail(string usernameOrAccountId)
...
+            if (JiraNameFieldCache == null)
             {
-                var nameField = (JValue)item.SelectToken("name");
-                if (nameField.Value.ToString().ToLower() == propertyName.ToLower())
-                {
-                    var idField = (JValue)item.SelectToken("id");
-                    customId = idField.Value.ToString();
-                }
+                var response = (JArray)_jiraServiceWrapper.RestClient.ExecuteRequestAsync(Method.GET, $"{JiraApiV2}/field").Result;
+                JiraNameFieldCache = response
```
Comment (madkoo) [inline src/WorkItemMigrator/JiraExport/JiraProvider.cs:401]: I dont follow the duplication here ?  from line 400 to 410. 
Why do you need the key as well ?

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs [hunk 4]
```diff
@@ -405,25 +395,40 @@ public string GetUserEmail(string usernameOrAccountId)
...
+        }
 
-                if (string.IsNullOrEmpty(customId))
+        private string GetItemFromFieldCache(string propertyName, ILookup<string,string> cache)
+        {
+            string customId = null;
+            var query = cache.Where(x => x.Key.Equals(propertyName.ToLower())).FirstOrDefault();
+            if (query != null)
+            {
+                customId = query.Count() > 0 ? query.First() : null;
```
Comment (madkoo) [inline src/WorkItemMigrator/JiraExport/JiraProvider.cs:428]: How do you know if the first one is correct if you get multiple hits in the grouping ?

**Review by madkoo — COMMENTED** [review]
Added comments, please check.

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs (hunk 3 above)
Comment (moerketh) [reply src/WorkItemMigrator/JiraExport/JiraProvider.cs:401]: My Jira instance does not return a "key" property on any of the fields. However, the existing code checks the "name" property first and if no match is found it checks the "key" property. Does that answer your question?

Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/JiraExport/JiraProvider.cs:401]: @moerketh

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs (hunk 4 above)
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/JiraExport/JiraProvider.cs:428]: This was the same behavior as we had before, in the case where we received multiple fields with the same key. We added a warning statement in can we have multiple cache hits

**Comment by moerketh** [comment]
Only just realized you are keeping 0 issues in SonarCloud, very nice. 
I raised PR #542 to attempt to fix the code smells that remain in this branch.

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs (hunk 4 above)
Comment (Alexander-Hjelm) [reply src/WorkItemMigrator/JiraExport/JiraProvider.cs]: @madkoo , @moerketh , what is the status here? :)

Comment (madkoo) [reply src/WorkItemMigrator/JiraExport/JiraProvider.cs]: Ok, let it be then.

File: src/WorkItemMigrator/JiraExport/JiraProvider.cs (hunk 3 above)
Comment (madkoo) [reply src/WorkItemMigrator/JiraExport/JiraProvider.cs:401]: I mean, I would move the KeyCache inside the if statement where we actually try to use it and there check if the key cache is null or not and then fill it.

**Review by madkoo — APPROVED** [review]
Looks good :)

**Comment by Alexander-Hjelm** [comment]
Cool! integration tests are running smoothly as well, merging when the checks complete

**Pull request merged**